metadata-ingestion/src/datahub/ingestion/source/redash.py

```diff
@@ -0,0 +1,462 @@
+    "SANKEY": DEFAULT_VISUALIZATION_TYPE,
+    "SUNBURST_SEQUENCE": DEFAULT_VISUALIZATION_TYPE,
+    "TABLE": ChartTypeClass.TABLE,
+    "WORD_CLOUD": DEFAULT_VISUALIZATION_TYPE,
+}
+
+
+class RedashConfig(ConfigModel):
+    # See the Redash API for details
+    # https://redash.io/help/user-guide/integrations-and-api/api
+    connect_uri: str = "http://localhost:5000"
+    api_key: str = "REDASH_API_KEY"
+    env: str = DEFAULT_ENV
+
+    # Optionals
+    dashboard_patterns: AllowDenyPattern = AllowDenyPattern.allow_all()
+    chart_patterns: AllowDenyPattern = AllowDenyPattern.allow_all()
+    skip_draft: bool = True
+    api_page_limit: int = sys.maxsize
+    # parse_table_names_from_sql: bool = False  # TODO: _get_upstream_lineage from SQL
+
+
+@dataclass
+class RedashSourceReport(SourceReport):
+    items_scanned: int = 0
+    filtered: List[str] = field(default_factory=list)
+
+    def report_topic_scanned(self, item: str) -> None:
+        self.items_scanned += 1
+
+    def report_dropped(self, item: str) -> None:
+        self.filtered.append(item)
+
+
+class RedashSource(Source):
+    config: RedashConfig
+    report: RedashSourceReport
+    platform = "redash"
+
+    def __hash__(self):
+        return id(self)
+
+    def __init__(self, ctx: PipelineContext, config: RedashConfig):
+        super().__init__(ctx)
+        self.config = config
+        self.report = RedashSourceReport()
+
+        # Handle trailing slash removal
+        self.config.connect_uri = self.config.connect_uri.strip("/")
+
+        self.client = Redash(self.config.connect_uri, self.config.api_key)
+        self.client.session.headers.update(
+            {
+                "Content-Type": "application/json",
+                "Accept": "application/json",
+            }
+        )
+
+        self.api_page_limit = self.config.api_page_limit or math.inf
+
+        # Test the connection
+        test_response = self.client._get(f"{self.config.connect_uri}/api")
+        if test_response.status_code == 200:
+            logger.info("Redash API connected succesfully")
+            pass
+
+            # # Only for getting data source types available in Redash
+            # # We use this as source for `DATA_SOURCE_TO_PLATFORM_MAP` variable
+            # data_source_types = self.client._get(f"/api/data_sources/types").json()
+            # for dst in data_source_types:
+            #     print(f'\'{dst["type"]}\'', ":", {"name": dst["name"], "platform": None, }, ",")
+
+    @classmethod
+    def create(cls, config_dict: dict, ctx: PipelineContext) -> Source:
+        config = RedashConfig.parse_obj(config_dict)
+        return cls(ctx, config)
+
+    @lru_cache(maxsize=None)
+    def _get_chart_data_source(self, data_source_id):
+        url = f"/api/data_sources/{data_source_id}"
+        resp = self.client._get(url).json()
+        return resp
+
+    @lru_cache(maxsize=None)
+    def _get_datasource_urn_from_data_source_id(self, data_source_id):
```

Comment:
@datascienceChris this is current approach on how to get chart sources

Reply:
Add some type annotations here

Reply:
already added type annotations